### PR TITLE
Switch ASSERT macros to new stream API

### PR DESCRIPTION
**Pull request by bejado**
There's still a handful of cases I need to manually switch over. In terms of review, maybe we can divide-and-conquer. The biggest risk is the formatting won't match the old printf style.

### CODE REVIEW

File: libs/bluevk/src/BlueVKDarwin.cpp [hunk 1]
```diff
@@ -31,10 +31,10 @@ bool loadLibrary() {
 #endif
 
     module = dlopen(dylibPath, RTLD_NOW | RTLD_LOCAL);
-    ASSERT_POSTCONDITION(module != nullptr,
-            "BlueVK is unable to load entry points: %s.\n"
-            "Install the LunarG SDK with 'System Global Installation' and reboot.\n",
-            dlerror());
+    FILAMENT_CHECK_POSTCONDITION(module != nullptr)
+            << "BlueVK is unable to load entry points: " << dlerror()
+            << ".\n"
```
Comment (poweifeng) [inline libs/bluevk/src/BlueVKDarwin.cpp:36]:
```suggestion
            << "BlueVK is unable to load entry points: " << dlerror() << ".\n"
```

File: libs/filabridge/src/SamplerInterfaceBlock.cpp [hunk 2]
```diff
@@ -93,8 +93,7 @@ SamplerInterfaceBlock::SamplerInterfaceBlock(Builder const& builder) noexcept
 const SamplerInterfaceBlock::SamplerInfo* SamplerInterfaceBlock::getSamplerInfo(
         std::string_view name) const {
     auto pos = mInfoMap.find(name);
-    ASSERT_PRECONDITION(pos != mInfoMap.end(), "sampler named \"%.*s\" not found",
-            name.size(), name.data());
+    FILAMENT_CHECK_PRECONDITION(pos != mInfoMap.end()) << "sampler named \"%.*s\" not found";
```
Comment (poweifeng) [inline libs/filabridge/src/SamplerInterfaceBlock.cpp:96]:
```suggestion
    FILAMENT_CHECK_PRECONDITION(pos != mInfoMap.end()) << "sampler named \"" << name << "\" not found";
```

File: filament/src/components/RenderableManager.cpp [hunk 3]
```diff
@@ -432,30 +439,32 @@ void RenderableManager::BuilderDetails::processBoneIndicesAndWights(Engine& engi
...
 
-    ASSERT_PRECONDITION(mImpl->mInstanceCount <= CONFIG_MAX_INSTANCES || !mImpl->mInstanceBuffer,
-            "instance count is %zu, but instance count is limited to CONFIG_MAX_INSTANCES (%zu) "
-            "instances when supplying transforms via an InstanceBuffer.",
-            mImpl->mInstanceCount,
-            CONFIG_MAX_INSTANCES);
+    FILAMENT_CHECK_PRECONDITION(
+            mImpl->mInstanceCount <= CONFIG_MAX_INSTANCES || !mImpl->mInstanceBuffer)
+            << "instance count is " << mImpl->mInstanceCount
+            << "u, but instance count is limited to CONFIG_MAX_INSTANCES (" << CONFIG_MAX_INSTANCES
```
Comment (poweifeng) [inline filament/src/components/RenderableManager.cpp:448]:
```suggestion
            << ", but instance count is limited to CONFIG_MAX_INSTANCES (" << CONFIG_MAX_INSTANCES
```

File: filament/src/components/RenderableManager.cpp [hunk 4]
```diff
@@ -432,30 +439,32 @@ void RenderableManager::BuilderDetails::processBoneIndicesAndWights(Engine& engi
...
+            << "u) "
+               "instances when supplying transforms via an InstanceBuffer.";
```
Comment (poweifeng) [inline filament/src/components/RenderableManager.cpp:450]:
```suggestion
            << ") instances when supplying transforms via an InstanceBuffer.";
```

File: filament/src/components/RenderableManager.cpp [hunk 5]
```diff
@@ -432,30 +439,32 @@ void RenderableManager::BuilderDetails::processBoneIndicesAndWights(Engine& engi
...
     if (mImpl->mInstanceBuffer) {
         size_t const bufferInstanceCount = mImpl->mInstanceBuffer->mInstanceCount;
-        ASSERT_PRECONDITION(mImpl->mInstanceCount <= bufferInstanceCount,
-                "instance count (%zu) must be less than or equal to the InstanceBuffer's instance "
-                "count "
-                "(%zu).",
-                mImpl->mInstanceCount, bufferInstanceCount);
+        FILAMENT_CHECK_PRECONDITION(mImpl->mInstanceCount <= bufferInstanceCount)
+                << "instance count (" << mImpl->mInstanceCount
+                << "u) must be less than or equal to the InstanceBuffer's instance "
```
Comment (poweifeng) [inline filament/src/components/RenderableManager.cpp:464]:
```suggestion
                << ") must be less than or equal to the InstanceBuffer's instance "
```

File: filament/src/components/RenderableManager.cpp [hunk 6]
```diff
@@ -432,30 +439,32 @@ void RenderableManager::BuilderDetails::processBoneIndicesAndWights(Engine& engi
...
+                   "count "
+                   "("
+                << bufferInstanceCount << "u).";
```
Comment (poweifeng) [inline filament/src/components/RenderableManager.cpp:467]:
```suggestion
                   "count (" << bufferInstanceCount << ").";
```

File: filament/backend/src/vulkan/platform/VulkanPlatformAndroidLinuxWindows.cpp [hunk 7]
```diff
@@ -123,8 +123,8 @@ VulkanPlatform::SurfaceBundle VulkanPlatform::createVkSurfaceKHR(void* nativeWin
         };
         VkResult const result = vkCreateAndroidSurfaceKHR(instance, &createInfo, VKALLOC,
                 (VkSurfaceKHR*) &surface);
-        ASSERT_POSTCONDITION(result == VK_SUCCESS, "vkCreateAndroidSurfaceKHR error.");
     #elif defined(__linux__) && defined(FILAMENT_SUPPORTS_WAYLAND)
+        FILAMENT_CHECK_POSTCONDITION(result == VK_SUCCESS) << "vkCreateAndroidSurfaceKHR error.";
```
[the remark this answers is not in the document]
Comment (bejado) [reply filament/backend/src/vulkan/platform/VulkanPlatformAndroidLinuxWindows.cpp:127]: Good catch, not sure what happened here.

File: filament/src/details/InstanceBuffer.cpp [hunk 8]
```diff
@@ -51,12 +51,12 @@ InstanceBuffer::Builder& InstanceBuffer::Builder::localTransforms(
...
-            "instanceCount is %zu, but instance count is limited to "
-            "Engine::getMaxAutomaticInstances() (%zu) instances when supplying transforms.",
-            mImpl->mInstanceCount,
-            engine.getMaxAutomaticInstances());
+    FILAMENT_CHECK_PRECONDITION(mImpl->mInstanceCount >= 1) << "instanceCount must be >= 1.";
+    FILAMENT_CHECK_PRECONDITION(mImpl->mInstanceCount <= engine.getMaxAutomaticInstances())
+            << "instanceCount is " << mImpl->mInstanceCount
+            << "u, but instance count is limited to "
+               "Engine::getMaxAutomaticInstances() ("
+            << engine.getMaxAutomaticInstances() << "u) instances when supplying transforms.";
```
Comment (poweifeng) [inline filament/src/details/InstanceBuffer.cpp:59]:
```suggestion
            << "instanceCount is " << mImpl->mInstanceCount
            << ", but instance count is limited to Engine::getMaxAutomaticInstances() ("
            << engine.getMaxAutomaticInstances() << ") instances when supplying transforms.";
```

File: filament/src/details/InstanceBuffer.cpp [hunk 9]
```diff
@@ -76,12 +76,11 @@ FInstanceBuffer::FInstanceBuffer(FEngine& engine, const Builder& builder) {
...
-            "setLocalTransforms overflow. InstanceBuffer has only %zu instances, but trying to set "
-            "%zu transforms at offset %zu.",
-            mInstanceCount,
-            count,
-            offset);
+    FILAMENT_CHECK_PRECONDITION(offset + count <= mInstanceCount)
+            << "setLocalTransforms overflow. InstanceBuffer has only " << mInstanceCount
+            << "u instances, but trying to set "
+               ""
+            << count << "u transforms at offset " << offset << "u.";
```
Comment (poweifeng) [inline filament/src/details/InstanceBuffer.cpp:83]:
```suggestion
            << "setLocalTransforms overflow. InstanceBuffer has only " << mInstanceCount
            << " instances, but trying to set " << count 
            << " transforms at offset " << offset << ".";
```

File: filament/src/details/Material.cpp [hunk 10]
```diff
@@ -91,23 +91,26 @@ static std::unique_ptr<MaterialParser> createParser(Backend backend,
...
-        ASSERT_PRECONDITION(materialResult != MaterialParser::ParseResult::ERROR_MISSING_BACKEND,
-                "the material was not built for any of the %s backend's supported shader "
-                "languages (%s)\n",
-                backendToString(backend), languageNames.c_str());
+        FILAMENT_CHECK_PRECONDITION(
+                materialResult != MaterialParser::ParseResult::ERROR_MISSING_BACKEND)
+                << "the material was not built for any of the " << backendToString(backend)
+                << " backend's supported shader "
+                   "languages ("
+                << languageNames.c_str() << ")\n";
```
Comment (poweifeng) [inline filament/src/details/Material.cpp:99]:
```suggestion
                << " backend's supported shader languages (" << languageNames.c_str() << ")\n";
```

File: filament/src/details/Material.cpp [hunk 11]
```diff
@@ -522,10 +525,11 @@ Program FMaterial::getProgramWithVariants(
...
 
-    ASSERT_POSTCONDITION(isNoop || (vsOK && !vsBuilder.empty()),
-            "The material '%s' has not been compiled to include the required "
-            "GLSL or SPIR-V chunks for the vertex shader (variant=0x%x, filtered=0x%x).",
-            mName.c_str(), variant.key, vertexVariant.key);
+    FILAMENT_CHECK_POSTCONDITION(isNoop || (vsOK && !vsBuilder.empty()))
+            << "The material '" << mName.c_str()
+            << "' has not been compiled to include the required "
+               "GLSL or SPIR-V chunks for the vertex shader (variant=0x"
+            << variant.key << ", filtered=0x" << vertexVariant.key << ").";
```
Comment (poweifeng) [inline filament/src/details/Material.cpp:532]:
```suggestion
               "GLSL or SPIR-V chunks for the vertex shader (variant=0x" << io::hex
            << variant.key << ", filtered=0x" << vertexVariant.key << io::dex << ").";
```

File: filament/src/details/Material.cpp [hunk 12]
```diff
@@ -536,10 +540,11 @@ Program FMaterial::getProgramWithVariants(
...
 
-    ASSERT_POSTCONDITION(isNoop || (fsOK && !fsBuilder.empty()),
-            "The material '%s' has not been compiled to include the required "
-            "GLSL or SPIR-V chunks for the fragment shader (variant=0x%x, filtered=0x%x).",
-            mName.c_str(), variant.key, fragmentVariant.key);
+    FILAMENT_CHECK_POSTCONDITION(isNoop || (fsOK && !fsBuilder.empty()))
+            << "The material '" << mName.c_str()
+            << "' has not been compiled to include the required "
+               "GLSL or SPIR-V chunks for the fragment shader (variant=0x"
+            << variant.key << ", filtered=0x" << fragmentVariant.key << ").";
```
Comment (poweifeng) [inline filament/src/details/Material.cpp:547]:
```suggestion
               "GLSL or SPIR-V chunks for the fragment shader (variant=0x" << io::hex
            << variant.key << ", filtered=0x" << fragmentVariant.key << io::dec << ").";
```

File: filament/src/details/MorphTargetBuffer.cpp [hunk 13]
```diff
@@ -148,15 +148,14 @@ void FMorphTargetBuffer::terminate(FEngine& engine) {
 
 void FMorphTargetBuffer::setPositionsAt(FEngine& engine, size_t targetIndex,
         math::float3 const* positions, size_t count, size_t offset) {
-
-    ASSERT_PRECONDITION(offset + count <= mVertexCount,
-            "MorphTargetBuffer (size=%lu) overflow (count=%u, offset=%u)",
-            (unsigned)mVertexCount, (unsigned)count, (unsigned)offset);
+    FILAMENT_CHECK_PRECONDITION(offset + count <= mVertexCount)
+            << "MorphTargetBuffer (size=" << (unsigned)mVertexCount
+            << "u) overflow (count=" << (unsigned)count << ", offset=" << (unsigned)offset << ")";
```
Comment (poweifeng) [inline filament/src/details/MorphTargetBuffer.cpp:153]:
```suggestion
            << "MorphTargetBuffer (size=" << (unsigned)mVertexCount
            << ") overflow (count=" << (unsigned)count << ", offset=" << (unsigned)offset << ")";
```

File: filament/src/details/MorphTargetBuffer.cpp [hunk 14]
```diff
@@ -172,15 +171,14 @@ void FMorphTargetBuffer::setPositionsAt(FEngine& engine, size_t targetIndex,
 
 void FMorphTargetBuffer::setPositionsAt(FEngine& engine, size_t targetIndex,
         math::float4 const* positions, size_t count, size_t offset) {
-
-    ASSERT_PRECONDITION(offset + count <= mVertexCount,
-            "MorphTargetBuffer (size=%lu) overflow (count=%u, offset=%u)",
-            (unsigned)mVertexCount, (unsigned)count, (unsigned)offset);
+    FILAMENT_CHECK_PRECONDITION(offset + count <= mVertexCount)
+            << "MorphTargetBuffer (size=" << (unsigned)mVertexCount
+            << "u) overflow (count=" << (unsigned)count << ", offset=" << (unsigned)offset << ")";
```
Comment (poweifeng) [inline filament/src/details/MorphTargetBuffer.cpp:176]:
```suggestion
            << "MorphTargetBuffer (size=" << (unsigned)mVertexCount
            << ") overflow (count=" << (unsigned)count << ", offset=" << (unsigned)offset << ")";
```

File: filament/src/details/MorphTargetBuffer.cpp [hunk 15]
```diff
@@ -195,15 +193,14 @@ void FMorphTargetBuffer::setPositionsAt(FEngine& engine, size_t targetIndex,
 
 void FMorphTargetBuffer::setTangentsAt(FEngine& engine, size_t targetIndex,
         math::short4 const* tangents, size_t count, size_t offset) {
-
-    ASSERT_PRECONDITION(offset + count <= mVertexCount,
-            "MorphTargetBuffer (size=%lu) overflow (count=%u, offset=%u)",
-            (unsigned)mVertexCount, (unsigned)count, (unsigned)offset);
+    FILAMENT_CHECK_PRECONDITION(offset + count <= mVertexCount)
+            << "MorphTargetBuffer (size=" << (unsigned)mVertexCount
+            << "u) overflow (count=" << (unsigned)count << ", offset=" << (unsigned)offset << ")";
```
Comment (poweifeng) [inline filament/src/details/MorphTargetBuffer.cpp:198]:
```suggestion
            << ") overflow (count=" << (unsigned)count << ", offset=" << (unsigned)offset << ")";
```

File: filament/src/details/SkinningBuffer.cpp [hunk 16]
```diff
@@ -96,20 +96,20 @@ void FSkinningBuffer::terminate(FEngine& engine) {
...
 void FSkinningBuffer::setBones(FEngine& engine,
         RenderableManager::Bone const* transforms, size_t count, size_t offset) {
-
-    ASSERT_PRECONDITION((offset + count) <= mBoneCount,
-            "SkinningBuffer (size=%lu) overflow (boneCount=%u, offset=%u)",
-            (unsigned)mBoneCount, (unsigned)count, (unsigned)offset);
+    FILAMENT_CHECK_PRECONDITION((offset + count) <= mBoneCount)
+            << "SkinningBuffer (size=" << (unsigned)mBoneCount
+            << "u) overflow (boneCount=" << (unsigned)count << ", offset=" << (unsigned)offset
+            << ")";
```
Comment (poweifeng) [inline filament/src/details/SkinningBuffer.cpp:102]:
```suggestion
            << ") overflow (boneCount=" << (unsigned)count << ", offset=" << (unsigned)offset
            << ")";
```

File: filament/src/details/SkinningBuffer.cpp [hunk 17]
```diff
@@ -96,20 +96,20 @@ void FSkinningBuffer::terminate(FEngine& engine) {
...
 void FSkinningBuffer::setBones(FEngine& engine,
         math::mat4f const* transforms, size_t count, size_t offset) {
-
-    ASSERT_PRECONDITION((offset + count) <= mBoneCount,
-            "SkinningBuffer (size=%lu) overflow (boneCount=%u, offset=%u)",
-            (unsigned)mBoneCount, (unsigned)count, (unsigned)offset);
+    FILAMENT_CHECK_PRECONDITION((offset + count) <= mBoneCount)
+            << "SkinningBuffer (size=" << (unsigned)mBoneCount
+            << "u) overflow (boneCount=" << (unsigned)count << ", offset=" << (unsigned)offset
+            << ")";
```
Comment (poweifeng) [inline filament/src/details/SkinningBuffer.cpp:112]:
```suggestion
            << ") overflow (boneCount=" << (unsigned)count << ", offset=" << (unsigned)offset
            << ")";
```

File: filament/src/details/Texture.cpp [hunk 18]
```diff
@@ -186,8 +187,8 @@ Texture* Texture::Builder::build(Engine& engine) {
     const bool imported = mImpl->mImportedId;
 
     #if defined(__EMSCRIPTEN__)
-    ASSERT_PRECONDITION(!swizzled, "WebGL does not support texture swizzling.");
     #endif
+    FILAMENT_CHECK_PRECONDITION(!swizzled) << "WebGL does not support texture swizzling.";
```
Comment (poweifeng) [inline filament/src/details/Texture.cpp:191]: this assertion has been moved

File: filament/src/details/Texture.cpp [hunk 19]
```diff
@@ -326,22 +328,24 @@ void FTexture::setImage(FEngine& engine, size_t level,
...
-        "buffer overflow: (size=%lu, stride=%lu, left=%u, top=%u) smaller than specified region "
-        "{{%u,%u,%u},{%u,%u,%u)}}",
-        size_t(p.size), size_t(p.stride), unsigned(p.left), unsigned(p.top),
-        unsigned(xoffset), unsigned(yoffset), unsigned(zoffset),
-        unsigned(width), unsigned(height), unsigned(depth));
+    FILAMENT_CHECK_PRECONDITION(bpp * p.left + bpr * p.top + bpl * (0 + depth) <= p.size)
+            << "buffer overflow: (size=" << size_t(p.size) << "u, stride=" << size_t(p.stride)
+            << "u, left=" << unsigned(p.left) << ", top=" << unsigned(p.top)
+            << ") smaller than specified region "
+               "{{"
```
Comment (poweifeng) [inline filament/src/details/Texture.cpp:346]:
```suggestion
            << "buffer overflow: (size=" << size_t(p.size) << ", stride=" << size_t(p.stride)
            << ", left=" << unsigned(p.left) << ", top=" << unsigned(p.top)
            << ") smaller than specified region "
               "{{"
```

File: filament/src/fg/Resource.cpp [hunk 20]
```diff
@@ -100,9 +100,9 @@ void ImportedRenderTarget::assertConnect(FrameGraphTexture::Usage u) {
                                     FrameGraphTexture::Usage::DEPTH_ATTACHMENT |
                                     FrameGraphTexture::Usage::STENCIL_ATTACHMENT;
 
-    ASSERT_PRECONDITION(none(u & ~ANY_ATTACHMENT),
-            "Imported render target resource \"%s\" can only be used as an attachment (usage=%s)",
-            name, utils::to_string(u).c_str());
+    FILAMENT_CHECK_PRECONDITION(none(u & ~ANY_ATTACHMENT))
+            << "Imported render target resource \"" << name
+            << "\" can only be used as an attachment (usage=" << utils::to_string(u).c_str() << ")";
```
Comment (poweifeng) [inline filament/src/fg/Resource.cpp:105]:
```suggestion
            << "\" can only be used as an attachment (usage=" << utils::to_string(u).c_str() << ')';
```

**Review by pixelflinger — APPROVED** [review]
lgtm, but I think we should check the conversion tool in for posterity.

**Pull request merged**